### PR TITLE
Caching figures generated by plot directive

### PR DESCRIPTION
**Pull request by dstansby**
## PR Summary
A rebase of https://github.com/matplotlib/matplotlib/pull/10149 onto current main, with conflicts solved, and the latest feedback on that PR taken into account. For context see the description of https://github.com/matplotlib/matplotlib/pull/10149.

## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by ksunden** [comment]
How does this interact with the changes to the plot directive introduced in #24205?

My initial reaction is that the `outname` change should be orthogonal, but the `plot_cache_dir` may conflict and/or be totally ignored.

File: doc/users/next_whats_new/sphinx_plot_preserve.rst [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+       import matplotlib.image as mpimg
+       import numpy as np
+       img = mpimg.imread('_static/stinkbug.png')
+       imgplot = plt.imshow(img)
+
+Without ``:outname:``, the figure generated above would normally be called,
+e.g. :file:`docfile3-4-01.png` or something equally mysterious. With
+``:outname:`` the figure generated will instead be named
+:file:`stinkbug_plot-01.png` or even :file:`stinkbug_plot.png`. This makes it
+easy to understand which output image is which and, more importantly, uniquely
```
Comment (tacaswell) [inline doc/users/next_whats_new/sphinx_plot_preserve.rst:22]: I am slightly skeptical of the first claim (that users can give things good names) and _very_ skeptical of the second as the name is not random, it is based on the name of the file (and the position + function of the directive in that file) so is structurally unique.  If users can control the names there is a high chance that they are going to collide.

If we want to make things save to cache like this we should give them names based on the hash of the source....

File: lib/matplotlib/sphinxext/plot_directive.py [hunk 2]
```diff
@@ -613,6 +635,12 @@ def render_figures(code, code_path, output_dir, output_base, context,
             for fmt, dpi in formats:
                 try:
                     figman.canvas.figure.savefig(img.filename(fmt), dpi=dpi)
+                    if config.plot_cache_dir and outname is not None:
```
Comment (tacaswell) [inline lib/matplotlib/sphinxext/plot_directive.py:638]: We do not actually use `outname` for anything other than checking it exists!

**Review by tacaswell — CHANGES_REQUESTED** [review]
I have some conceptual concerns:
 - name are already (structurally) unique
 - do not trust users to give things unique names (although it does check that)
 - the cache is not invalidated if the code in the directive changes

and an implementation concern (we do not seem to actually use `outname` in the code!)

I think this is a good idea, but probably should cache the files based on a hash of the source rather than a user supplied name.

**Comment by dstansby** [comment]
Sorry for doing a 'blind' rebase and not really thinking this through myself, but thanks for re-reviewing! Just thinking about the key for the cache, hasing the source code is potentially no though good because changes in Matplotlib internals (e.g. a change in default style, a bugfix to a plotting method) would lead to a different image with the same source code.

So I think we either have to:

1. Abandon this
2. Use a hash of the source code of the key, with some suitable user warning that any cached images will only ever be automatically updated if the source code is updated.

I'm torn between these two options, but am erring towards number 1. due to the potential of number 2. to cause comfusion to users at some point down the line. Thoughts?

**Comment by tacaswell** [comment]
> Sorry for doing a 'blind' rebase and not really thinking this through myself, but thanks for re-reviewing! 

I went back to the original PR where I also did a blind rebase and am to blame for at least some of these issues!

For the key we could start the hash with `matplotlib.__version__`, but for developers that would quickly lead to an ~unbounded cache as we put the git sha in the version and would only move the problem out a level to any plot directive that uses Matplotlib via a library (e.g. seaborn) having exactly the same problem but not participating in the hash (and it seems this would be general to any library used in the code...).

So given the caching problem (nb: there are two hard problems in computer science: naming things, cache invalidation, and off-by-one bugs) I'm inclined to go with option 1 as well.

**Comment by dstansby** [comment]
In that case I'll close. @r-barnes thanks a lot for opening this PR originally - we decided not to carry it forward because of the difficulties of caching the images. See the comment above this one for more context. I think we would be open to re-visiting this, but that would require a well thought out strategy for caching the images in an issue before opening a pull request.

**Pull request closed without merging**